socketio/src/client/reconnect.rs

```diff
@@ -0,0 +1,239 @@
+            backoff: ExponentialBackoff::default(),
+        };
+        s.poll_callback();
+
+        Ok(s)
+    }
+
+    pub fn emit<E, D>(&self, event: E, data: D) -> Result<()>
+    where
+        E: Into<Event>,
+        D: Into<Payload>,
+    {
+        let client = self.client.read()?;
+        // TODO(#230): like js client, buffer emit, resend after reconnect
+        client.emit(event, data)
+    }
+
+    pub fn emit_with_ack<F, E, D>(
+        &self,
+        event: E,
+        data: D,
+        timeout: Duration,
+        callback: F,
+    ) -> Result<()>
+    where
+        F: for<'a> FnMut(Payload, Client) + 'static + Sync + Send,
+        E: Into<Event>,
+        D: Into<Payload>,
+    {
+        let client = self.client.read()?;
+        // TODO(#230): like js client, buffer emit, resend after reconnect
+        client.emit_with_ack(event, data, timeout, callback)
+    }
+
+    pub fn disconnect(&self) -> Result<()> {
+        let client = self.client.read()?;
+        client.disconnect()
+    }
+
+    fn reconnect(&mut self) -> Result<()> {
+        let mut reconnect_attempts = 0;
+        if self.builder.reconnect {
+            loop {
+                // 0 means infinity
+                if self.builder.max_reconnect_attempts != 0
+                    && reconnect_attempts > self.builder.max_reconnect_attempts
+                {
+                    break;
+                }
+                reconnect_attempts += 1;
+
+                if let Some(backoff) = self.backoff.next_backoff() {
+                    std::thread::sleep(backoff);
+                }
+
+                if self.do_reconnect().is_ok() {
+                    self.poll_callback();
+                    break;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn do_reconnect(&self) -> Result<()> {
+        let new_client = self.builder.clone().connect_manual()?;
+        let mut client = self.client.write()?;
+        *client = new_client;
+
+        Ok(())
+    }
+
+    pub(crate) fn iter(&self) -> Iter {
+        Iter {
+            socket: self.client.clone(),
+        }
+    }
+
+    fn poll_callback(&self) {
+        let mut self_clone = self.clone();
+        // Use thread to consume items in iterator in order to call callbacks
+        std::thread::spawn(move || {
+            // tries to restart a poll cycle whenever a 'normal' error occurs,
+            // it just panics on network errors, in case the poll cycle returned
+            // `Result::Ok`, the server receives a close frame so it's safe to
+            // terminate
+            for packet in self_clone.iter() {
+                if let e @ Err(Error::IncompleteResponseFromEngineIo(_)) = packet {
+                    //TODO: 0.3.X handle errors
+                    //TODO: logging error
+                    let _ = self_clone.disconnect();
+                    break;
+                }
+            }
+            self_clone.reconnect();
```

Comment:
iterator has the rwlock, we can continue to call next on the iterator.
```suggestion
                    //TODO: logging error
                    let _ = self_clone.disconnect();
                    self_clone.reconnect()
                }
            }
```